### PR TITLE
feat(authentication): teams & roles initial support

### PR DESCRIPTION
**Pull request by kkopanidis**
Limitations:
- Cannot set new permissions for members
- <s>Doesn't send emails for invitations (email template has been added though)</s>
- Only team owners can invite new members
- <s>Creation of new teams is not currently supported</s>

**What kind of change does this PR introduce?** 

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `main` branch
- [x] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature 

**Other information:**

### CODE REVIEW

File: modules/authentication/src/admin/team.ts [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        },
+        name: 'CreateTeam',
+        description: 'Creates a new team',
+      },
+      new ConduitRouteReturnDefinition('Team', Team.name),
+      this.createTeam.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:60]: `GetTeams` getter is declared twice. Meant to implement `GetTeam`?

File: modules/authentication/src/admin/team.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        path: '/teams/members/:id',
+        action: ConduitRouteActions.GET,
+        urlParams: {
+          id: ConduitObjectId.Required,
+          skip: ConduitNumber.Optional,
+          limit: ConduitNumber.Optional,
+          sort: ConduitString.Optional,
+          search: ConduitString.Optional,
+        },
+        name: 'GetTeamMembers',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:88]: This is `GetTeamMember`.
Description needs to be updated.

File: modules/authentication/src/admin/team.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        teams: [Team.name],
+        count: ConduitNumber.Required,
+      }),
+      this.getTeams.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams/members/:id',
+        action: ConduitRouteActions.GET,
+        urlParams: {
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:81]: Most of these are `queryParams`

File: modules/authentication/src/admin/team.ts [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+
+export class TeamsAdmin {
+  constructor(private readonly grpcSdk: ConduitGrpcSdk) {}
+
+  declareRoutes(routingManager: RoutingManager) {
+    routingManager.route(
+      {
+        path: '/teams',
+        action: ConduitRouteActions.GET,
+        urlParams: {
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:28]: These are `queryParams`

File: modules/authentication/src/admin/team.ts [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+        description: 'Creates a new team',
+      },
+      new ConduitRouteReturnDefinition('Team', Team.name),
+      this.createTeam.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams',
+        action: ConduitRouteActions.GET,
+        urlParams: {
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:62]: These are `queryParams`

File: modules/authentication/src/admin/team.ts (hunk 2 above)
Comment (kon14) [reply modules/authentication/src/admin/team.ts:88]: We should probably add a complementary `GetTeamMembers`.

File: modules/authentication/src/handlers/local.ts [hunk 6]
```diff
@@ -19,6 +19,8 @@ import { Token, User } from '../models';
 import { status } from '@grpc/grpc-js';
 import { IAuthenticationStrategy } from '../interfaces/AuthenticationStrategy';
 import { TokenProvider } from './tokenProvider';
+import { Team } from '../models/Team.schema';
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:22]: Unused `Team` import.

File: modules/authentication/src/handlers/team.ts [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+        },
+        action: ConduitRouteActions.DELETE,
+        middlewares: ['authMiddleware'],
+      },
+      new ConduitRouteReturnDefinition('DeleteTeam', 'String'),
+      this.deleteTeam.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/team/members/:team',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:406]: This should be `/teams/:team/members`

File: modules/authentication/src/handlers/team.ts [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+        action: ConduitRouteActions.POST,
+        middlewares: ['authMiddleware'],
+      },
+      new ConduitRouteReturnDefinition(Team.name),
+      this.createTeam.bind(this),
+    );
+
+    routingManager.route(
+      {
+        path: '/team/members',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:379]: This should be `teams/:id/members`, with team id being a `urlParam`

File: modules/authentication/src/admin/team.ts [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+      },
+      new ConduitRouteReturnDefinition('GetTeams', {
+        teams: [Team.name],
+        count: ConduitNumber.Required,
+      }),
+      this.getTeams.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams/members/:id',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:79]: This should be `/teams/:id/members`

File: modules/authentication/src/admin/team.ts [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+      new ConduitRouteReturnDefinition('GetTeamMembers', {
+        members: [User.name],
+        count: ConduitNumber.Required,
+      }),
+      this.getTeamMembers.bind(this),
+    );
+
+    routingManager.route(
+      {
+        path: '/teams/members/:id',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:100]: This should be `/teams/:id/members`

File: modules/authentication/src/admin/team.ts [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        },
+        name: 'AddTeamMembers',
+        description: 'Add users as members to a team',
+      },
+      new ConduitRouteReturnDefinition('AddTeamMembers', 'String'),
+      this.addTeamMembers.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams/members/:id',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:119]: This should be `/teams/:id/members`

File: modules/authentication/src/admin/team.ts [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+        },
+        name: 'RemoveTeamMembers',
+        description: 'Remove members from a team',
+      },
+      new ConduitRouteReturnDefinition('RemoveTeamMembers', 'String'),
+      this.removeTeamMembers.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams/members/:id',
```
Comment (kon14) [inline modules/authentication/src/admin/team.ts:138]: This should be `/teams/:id/members`

File: modules/authentication/src/handlers/team.ts [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+        relation: role,
+      });
+    }
+    return 'ok';
+  }
+
+  declareRoutes(routingManager: RoutingManager): void {
+    routingManager.route(
+      {
+        path: '/team',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:364]: This should be `/teams`

File: modules/authentication/src/handlers/team.ts [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+        },
+        action: ConduitRouteActions.DELETE,
+        middlewares: ['authMiddleware'],
+      },
+      new ConduitRouteReturnDefinition('DeleteTeamMembersResponse', 'String'),
+      this.removeTeamMembers.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/team/:id',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:393]: This should be `/teams/:id`

File: modules/authentication/src/handlers/team.ts [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+      },
+      new ConduitRouteReturnDefinition('GetTeamMembers', {
+        members: [User.name],
+        count: ConduitNumber.Required,
+      }),
+      this.getTeamMembers.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/teams/members/:team',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:422]: This should be `/teams/:team/members`

File: modules/authentication/src/handlers/team.ts [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+        },
+        name: 'ChangeMemberRole',
+        description: 'Changes the roles of members in a team',
+      },
+      new ConduitRouteReturnDefinition('ChangeMemberRole', 'String'),
+      this.modifyRoles.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/team/invite/accept',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:442]: This should be `/teams/invite/accept`

File: modules/authentication/src/handlers/team.ts [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+        },
+        action: ConduitRouteActions.GET,
+        middlewares: ['authMiddleware'],
+      },
+      new ConduitRouteReturnDefinition('InviteAccepted', 'String'),
+      this.acceptInvite.bind(this),
+    );
+    routingManager.route(
+      {
+        path: '/team/invite',
```
Comment (kon14) [inline modules/authentication/src/handlers/team.ts:455]: This should be `/teams/invite`

**Review by kon14 — APPROVED** [review]

**Review by sdimitris — APPROVED** [review]

**Review by kon14 — APPROVED** [review]

**Pull request merged**